qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp

```diff
@@ -0,0 +1,358 @@
+
+void ALCDataLoadingModel::setLogs(const MatrixWorkspace_sptr &ws) {
+
+  std::vector<std::string> logs;
+
+  const auto &properties = ws->run().getProperties();
+  std::transform(properties.cbegin(), properties.cend(), std::back_inserter(logs),
+                 [](const auto &property) { return property->name(); });
+
+  // sort alphabetically
+  // cannot use standard sort alone as some logs are capitalised and some are
+  // not
+  std::sort(logs.begin(), logs.end(), [](const auto &log1, const auto &log2) {
+    // compare logs char by char and return pair of non-equal elements
+    const auto result =
+        std::mismatch(log1.cbegin(), log1.cend(), log2.cbegin(), log2.cend(),
+                      [](const auto &lhs, const auto &rhs) { return std::tolower(lhs) == std::tolower(rhs); });
+    // compare the two elements to decide which log should go first
+    return result.second != log2.cend() &&
+           (result.first == log1.cend() || std::tolower(*result.first) < std::tolower(*result.second));
+  });
+  m_logs = logs;
+}
+
+void ALCDataLoadingModel::setPeriods(const Workspace_sptr &loadedWs) {
+
+  size_t numPeriods = MuonAnalysisHelper::numPeriods(loadedWs);
+  std::vector<std::string> periods;
+  for (size_t i = 0; i < numPeriods; i++) {
+    std::stringstream buffer;
+    buffer << i + 1;
+    periods.emplace_back(buffer.str());
+  }
+  m_periods = periods;
+}
+MatrixWorkspace_sptr ALCDataLoadingModel::exportWorkspace() {
+  if (m_loadedData)
+    return std::const_pointer_cast<MatrixWorkspace>(m_loadedData);
+  return MatrixWorkspace_sptr{};
+}
+
+void ALCDataLoadingModel::setWsForMuonInfo(const std::string &filename) {
+
+  IAlgorithm_sptr loadAlg = AlgorithmManager::Instance().create("Load");
+  loadAlg->setChild(true); // Don't want workspaces in the ADS
+
+  // We need logs only but we have to use Load
+  // (can't use LoadMuonLogs as not all the logs would be
+  // loaded), so we load the minimum amount of data, i.e., one spectrum
+  loadAlg->setProperty("Filename", filename);
+  loadAlg->setPropertyValue("SpectrumMin", "1");
+  loadAlg->setPropertyValue("SpectrumMax", "1");
+  loadAlg->setPropertyValue("OutputWorkspace", "__NotUsed");
+  loadAlg->execute();
+
+  Workspace_sptr loadedWs = loadAlg->getProperty("OutputWorkspace");
+  setPeriods(loadedWs);
+
+  double firstGoodData = loadAlg->getProperty("FirstGoodData");
+  double timeZero = loadAlg->getProperty("TimeZero");
+  m_minTime = firstGoodData - timeZero;
+
+  m_wsForInfo = MuonAnalysisHelper::firstPeriod(loadedWs);
+
+  setLogs(m_wsForInfo);
+  // Update number of detectors for this new first run
+  m_numDetectors = m_wsForInfo->getInstrument()->getNumberDetectors();
+}
+
+/**
+ * Load new data and update the view accordingly
+ * @param files :: [input] range of files (user-specified or auto generated)
+ */
+void ALCDataLoadingModel::load(const IALCDataLoadingView *view) {
```

Comment:
@robertapplin I just remembered that I forgot to ask about this line, here the view is passed to the function from the model so I can get the inputs from the view. I avoided this in other model methods but this one requires quite a few of inputs. I don't know if there is a better way of doing this?

Reply:
Whoops I must have missed this! It would be preferred to pass each argument into the model `load` function individually so that the Model doesn't know about the view. Here the Model is communicating directly with the View instead of via the presenter

An alternative might be to use a std::tuple multiple items of data from the view to the model

Reply:
Thanks for pointing this out, I think its best to sort this out before we merge

Reply:
Okay, nice!